モデルデータの処理を実装

yasu307 opened the pull request:
## 変更内容
- todo, categoryのモデルを作成 [d1cdb3b](https://github.com/yasu307/todo-app-angular/pull/1/commits/d1cdb3bd3aabe662369c65a285198d4c8cab88ae)
- モデルを取得するServiceを作成 [bfa671e](https://github.com/yasu307/todo-app-angular/pull/1/commits/bfa671e6e01c6a8be4a8b08f18bd9a89811102c2)
- バックエンドAPIをシミュレートするin-memory-dataを作成 [167011d](https://github.com/yasu307/todo-app-angular/pull/1/commits/167011dfd92f894e8193c12beb1bdd9fcf576743)
- モデルの細かい変更 [b628e72](https://github.com/yasu307/todo-app-angular/pull/1/commits/b628e72a515e853e4418d91d325bf45f4597212d)
  - モデルの時間を表すプロパティをDate型に変更
  - stringをundefined許容に変更
- in-memory-dataを追加 [b628e72](https://github.com/yasu307/todo-app-angular/pull/1/commits/b628e72a515e853e4418d91d325bf45f4597212d)
- ファイル構成をフィーチャーモジュールに変更 [c21c228](https://github.com/yasu307/todo-app-angular/pull/1/commits/c21c22810cee9ac5363502b349b4db20dc8a538a)
- モデルではEnum型のプロパティをin-memory-dataではnumber型で表すように変更 [afcefe9](https://github.com/yasu307/todo-app-angular/pull/1/commits/afcefe9372d986cc88f1e21fb393a802c51ee3a5)
  - この変更に伴い、Serviceのgetメソッドに変換処理を追加

## 画面変更
### todo一覧画面(追加)
path : todo/list 
変更内容
- todo一覧を表示するように
<img width="1792" alt="angular1 todo:list" src="https://user-images.githubusercontent.com/29055153/173327787-fbe9edec-ad9a-4dd4-83eb-5a09e9a2132c.png">

### カテゴリ一覧画面(追加)
path : category/list 
変更内容
- カテゴリ一覧を表示するように
<img width="1792" alt="angular1 category:list" src="https://user-images.githubusercontent.com/29055153/173327862-6dee7285-cbce-4120-bfc6-a4bdeecd4680.png">

## 懸念点
- TypeScriptのEnum型は安全でないという記事を見たので、記事を参考にEnumStateとEnumColorという独自の型を作成しました。もっと他にいい実装方法はあるでしょうか？ 
- バックエンドAPIから取得するデータに独自の型を指定できないという想定でin-memory-dataを作成しましたが、この想定はあっているでしょうか？
- 上の想定からObservableのmapオペレーターを用いて変換を行いましたが、少しややこしい気もします。何かもっといい実装方法はあるでしょうか？調べたところインターセプタで変換することも可能だと思いました。
- フィーチャーモジュールに倣い自分なりにファイル構成をしてみましたがいかがでしょうか？Serviceなどはmodelファイルや独自ファイルに切り分けるてもいいかもしれないと思っています。

yasu307 committed:
create Models and add these component and router

yasu307 committed:
create service of todo and category

yasu307 committed:
create in-memory-data and change get method

yasu307 committed:
Change models and in-memory-data

モデルの時間を表すプロパティをDate型に変更
stringをundefined許容に変更
in-memory-dataに格納するデータを追加

yasu307 committed:
change file organization to use feature module

yasu307 committed:
Change in-memory-data and get method

in-memory-dataに格納するデータをバックエンドAPIから送られてくるデータに近づけた.
具体的には、Enumとして扱っているデータをnumberに変更した。
それに伴い、ServiceのgetメソッドにてEnumとして扱っているデータをnumberからEnum型に変換する処理を加えた。

h-oikawa-jp commented on src/app/models/category.ts:
```diff
@@ -0,0 +1,60 @@
+  FUCHSIA: new Color(6, 255, 102, 255),
+} as const;
+
+export type EnumColor = typeof EnumColor[keyof typeof EnumColor];
```
> TypeScriptのEnum型は安全でないという記事を見たので、記事を参考にEnumStateとEnumColorという独自の型を作成しました。もっと他にいい実装方法はあるでしょうか？

この安全でないというのはこの辺の記事ですかね？

https://www.kabuku.co.jp/developers/good-bye-typescript-enum

上記記事の通りTypeScript界隈ではEnumの代わりにUnion型が推奨されてますので、
基本的には @yasu307 さんの実装で問題ないかと思います。

(※そもそもTypeScriptのEnumは数値か文字列でしか定義出来ないので、
`class Color`などのように独自のクラスのオブジェクトを列挙値としたい場合はEnumは最初から候補に上がらないのではと。)

yasu307 replied:
> この安全でないというのはこの辺の記事ですかね？
> 
> https://www.kabuku.co.jp/developers/good-bye-typescript-enum

そのような記事です。

> (※そもそもTypeScriptのEnumは数値か文字列でしか定義出来ないので、
> class Colorなどのように独自のクラスのオブジェクトを列挙値としたい場合はEnumは最初から候補に上がらないのではと。)

そもそもオブジェクトを列挙値としたい時はEnumは使えないのですね。IxiaSのEnumは自由度が高かったので勘違いをしていました。

h-oikawa-jp commented on src/app/todo/todo.service.ts:
```diff
@@ -0,0 +1,28 @@
+  // 全てのtodoを取得する
+  // バックエンドAPIから受け取ったデータをTodo型に変換した結果を返す
+  getTodos(): Observable<Todo[]> {
+    return this.http.get<Todo[]>(this.todosUrl).pipe(
```
> バックエンドAPIから取得するデータに独自の型を指定できないという想定でin-memory-dataを作成しましたが、この想定はあっているでしょうか？

すみません、自分が質問の内容を理解しきれていないかも知れませんが、
そのAPIのレスポンスに独自の型を指定しているのが↑の `.get<Todo[]>` の部分という認識なのですが、違いますかね？

また、それとin-memory-dataを使用するという事への因果関係がイマイチ分からず・・・；

yasu307 replied:
すみません。勘違いをしていました。
オブジェクトリテラルを作成する際にオブジェクトを入れ子にするという発想が抜けていました。そのためin-memory-dataで作成するオブジェクトでTodoクラスの中のstateやCategoryクラスの中のcolorを表現できないと思っていました。

h-oikawa-jp commented on src/app/todo/todo.service.ts:
```diff
@@ -0,0 +1,28 @@
+  // バックエンドAPIから受け取ったデータをTodo型に変換した結果を返す
+  getTodos(): Observable<Todo[]> {
+    return this.http.get<Todo[]>(this.todosUrl).pipe(
+      // バックエンドAPIはstateプロパティにState.codeが格納されているため変換する
+      map(
+        (todos: Todo[]) =>  {
+          for (let todo of todos) codeToEnumState(todo)
+          return todos
+        }
+      )
```
> 上の想定からObservableのmapオペレーターを用いて変換を行いましたが、少しややこしい気もします。何かもっといい実装方法はあるでしょうか？調べたところインターセプタで変換することも可能だと思いました。

そもそもAPIからの戻りであるJSONとしてのTodoの構造は

```js
{
  id: number,
  ...
  state: number,
  ...
}
```

ですが、
マッピングしようとしているTodo interfaceの構造体としては

```js
{
  id: number,
  ...
  state: {
    code: number,
    name: string,
  }, // class State の構造体
  ...
}
```

です。

見ての通り構造が違うので両者をマッピングするにはどこかで変換を掛けてやらないといけません。

おっしゃる通り `angular/common/http` の[HttpInterceptor](https://angular.io/api/common/http/HttpInterceptor)でも実装出来るかと思いますが、
結局のところ必要な変換処理をHttpClientからレスポンスを受け取る前にするか後にするかの違いだけで、
個人的にはややこしさという意味では特に変わらないのではと感じます。


もしフロントでこのマッピング処理を行う事に違和感を覚えるなら、
そもそもフロントでの表示名(state.name)も含めてサーバから返してやって、
APIのI/FとフロントでのTodoモデルの構造を同じにしてやるというのを1つの修正案として挙げておきます。
(課題としては今の実装でも構いません)

yasu307 replied:
> もしフロントでこのマッピング処理を行う事に違和感を覚えるなら、
> そもそもフロントでの表示名(state.name)も含めてサーバから返してやって、
> APIのI/FとフロントでのTodoモデルの構造を同じにしてやるというのを1つの修正案として挙げておきます。

修正案ありがとうございます。こちらの修正を適用した結果が[e51e5ad](https://github.com/yasu307/todo-app-angular/pull/1/commits/e51e5ad2e23c3f5c8fc7288a847e740347206e55)です。ご確認よろしくお願いいたします。

修正内容は以下です。
- in-memory-dataのTodo.stateとCategory.colorの値を必要な情報が全て格納されるように変更。
- Todo.stateとCategory.colorの型定義を変更
  - バックエンドから取得したオブジェクトにはプロパティしか持たないためメソッドを削除
  - それに伴い、classからinterfaceに変更
- EnumColor, EnumStateはバックエンドから取得したプロパティとは関係がなくなったため、colorとstateのデフォルトの選択肢に変更
- 以上の変更によりgetメソッドのmapオペレーターは必要なくなったため削除

yasu307 committed:
Change in-memory-data and Models. And delete map operator of get method.

in-memory-dataのTodo.stateとCategory.colorの値を変更
- 必要な情報を全て格納しておく
Todo.stateとCategory.colorの型定義を変更
- バックエンドから取得したオブジェクトにはプロパティしか持たないためメソッドを削除
ーーそれに伴い、classからinterfaceに変更
EnumColor, EnumStateはバックエンドから取得したプロパティとは関係がなくなったため、Angular内で使用する選択肢とした
以上の変更によりgetメソッドのmapオペレーターは必要なくなったため削除
